my-mini-compiler/phase1-w25/src/lexer/lexer.c

```
@@ -164,25 +83,72 @@ int main(int argc, char* argv[]) {
+    file_buffer = (char *)malloc(file_size + 1);
+    if (!file_buffer) {
+        fprintf(stderr, "file is null\n");
+        free(file_buffer);
+        fclose(fp);
+        return 1;
+    }
+    memset(file_buffer, 0, file_size + 1);  // initialize file_buffer to null
+    size_t bytes_read = fread(file_buffer, 1, file_size, fp);
 
     if (bytes_read != file_size) {
         fprintf(stderr, "Could not read the whole file\n");
-        free(input_buffer);
         fclose(fp);
+        free(file_buffer);
         return 1;
     }
+    fclose(fp);
 
-    int position = 0;
-    Token token;
-
-    printf("Analyzing input:\n%s\n\n", input_buffer);
+    printf("Analyzing input:\n%s\n\n", file_buffer);
+    
+    regex_t regex[TOKEN_COUNT];
+    regmatch_t pmatch[TOKEN_COUNT][1];
+
+    bool matches[TOKEN_COUNT] = { false };
+    bool match, any_match;
+    Token cur_token = {TOKEN_NONE, "", current_line, ERROR_NONE};
+    TokenType cur_match;
+    int cur_longest;
+    size_t cur_position = 0;
+    
+    char* input_buffer = file_buffer;
 
+    compile_patterns(regex);
     do {
-        token = get_next_token(input_buffer, &position);
-        print_token(token);
-    } while (token.type != TOKEN_EOF);
-
+        any_match = false;
+        for (int i = 0; i < TOKEN_COUNT; ++i) {
+            match = !regexec(&regex[i], input_buffer, 1, pmatch[i], 0);  // Invert logic on return value
+            matches[i] = match;
+            any_match |= match;
+        }
+        if (any_match) {
+            cur_match = get_longest_match(pmatch, matches, &cur_longest);
+            if (cur_match == TOKEN_NEWLINE) {  // really don't like doing this here
+                ++current_line;
+            }
+            cur_token.type = cur_match;
+            cur_token.line = current_line;
+            cur_token.error = ERROR_NONE;
+            memcpy(cur_token.lexeme, input_buffer, cur_longest);  // huge bug if identifier is longer than 100 chars
```

Can we throw an error here? Or check to see if it is more than 100 and set the error property for the token to "TOO_LONG (or something along those lines)?